### PR TITLE
CUDA/ROCm shared interface

### PR DESCRIPTION
**Pull request by ekuznetsov139**
This PR adds several macros and typedefs to assist with writing kernels that seamlessly work with both CUDA and ROCm.

### CODE REVIEW

File: tensorflow/core/util/gpu_device_functions.h [hunk 1]
```diff
@@ -878,6 +931,28 @@ __device__ inline std::complex<double> operator/(
 }
 #endif  // GOOGLE_CUDA
 
+namespace functor {
+// ROCm hcc(clang) has severe difficulties dealing with std::complex directly due to a header issue.
+// This template assists in casting std::complex into the corresponding internal ROCm types.
+template <class T> 
+struct CpxMapT {
```
Comment (deven-amd) [inline tensorflow/core/util/gpu_device_functions.h]: `MAP_STD_COMPLEX_TO_HIP_COMPLEX` ?

**Comment by ekuznetsov139** [comment]
Not to this PR. I have a few other PRs lined up that depend on this one and #815

File: tensorflow/core/util/gpu_device_functions.h (hunk 1 above)
Comment (ekuznetsov139) [reply tensorflow/core/util/gpu_device_functions.h]: Too long I think?

Comment (deven-amd) [reply tensorflow/core/util/gpu_device_functions.h]: your call, but a more descriptive version than the current incarnation would be better. 
this is admittedly nit-picking :)

File: tensorflow/core/util/gpu_device_functions.h [hunk 2]
```diff
@@ -33,12 +35,63 @@ limitations under the License.
...
+#define gpuEventDestroy hipEventDestroy
+#define gpuEventSynchronize hipEventSynchronize
+#define gpuEventCreate hipEventCreate
+#define gpuEventCreateWithFlags hipEventCreateWithFlags
+#define gpuEventDisableTiming hipEventDisableTiming 
+typedef hipStream_t gpuStream_t;
+typedef hipEvent_t gpuEvent_t;
+typedef int cudaError;
+typedef int cudaError_t;
+static std::string cudaGetErrorString(int err) { return std::to_string(err); }
```
Comment (deven-amd) [inline tensorflow/core/util/gpu_device_functions.h]: wondering if we can figure out a way to cleanly share these instead

https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/stream_executor/gpu/gpu_types.h

https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/stream_executor/gpu/gpu_helpers.h

Comment (ekuznetsov139) [reply tensorflow/core/util/gpu_device_functions.h]: Defer this until after I got the rest of #798 merged perhaps?

**Comment by ROCm-Apps-Test** [comment]
Can one of the admins verify this patch?

**Comment by ekuznetsov139** [comment]
Submitted upstream as #36641

**Pull request merged**